### PR TITLE
changed mapping

### PR DESCRIPTION
**Pull request by spirinamayya**
Changed Id mapping

### CODE REVIEW

File: rectools/models/sasrec.py [hunk 1]
```diff
@@ -37,52 +38,56 @@ def __init__(
...
             torch.use_deterministic_algorithms(True)
-            seed_everything(self.random_state, workers=True)
+            seed_everything(random_state, workers=True)
 
-    def fit(
+    def _fit(
         self,
         dataset: RecDataset,
-    ):
+        *args: tp.Any,
```
Comment (blondered) [inline rectools/models/sasrec.py]: Здесь указываем уже только конкретные параметры, которая наша модель использует в _fit. Поскольку нет ничего кроме датасета - дропаем *args и **kwargs
Пример: https://github.com/spirinamayya/RecTools/blob/5faf9b9fa1d2304c6ab1ea20fe9916008b33ef0c/rectools/models/ease.py#L65

File: rectools/models/sasrec.py [hunk 2]
```diff
@@ -37,52 +38,56 @@ def __init__(
...
+        """TODO"""
         user_item_interactions = dataset.get_raw_interactions()
 
-        user_item_interactions[Columns.User] = user_item_interactions[Columns.User].astype(str)
-        user_item_interactions[Columns.Item] = user_item_interactions[Columns.Item].astype(str)
-
         users = user_item_interactions[Columns.User].value_counts()
         users = users[users >= 2]
         users = users.index.to_list()
         user_item_interactions = user_item_interactions[(user_item_interactions[Columns.User].isin(users))]
```
Comment (blondered) [inline rectools/models/sasrec.py]: Можно не вытаскивать индекс заранее. Просто `isin(users.index)`

File: rectools/models/sasrec.py [hunk 3]
```diff
@@ -140,46 +139,59 @@ def recommend(
...
-    weights: List[List[float]]
+class SASRecTargetInput:
+    """TODO"""
+
+    next_watch: torch.LongTensor
+    weights: torch.FloatTensor
 
 
 class SequenceTaskConverter:
     """Convert database data to use in particular model"""
```
Comment (blondered) [inline rectools/models/sasrec.py:162]: Передавать в `__init__` мощные данные (здесь `user_item_interactions`) и делать их обработку, чтобы запомнить только результат преобразований - это грязно. Для таких целей мы используем `construct` (пример здесь: https://github.com/spirinamayya/RecTools/blob/5faf9b9fa1d2304c6ab1ea20fe9916008b33ef0c/rectools/dataset/dataset.py#L95)

Но конкретно тут нам это не нужно)
В _fit основной модели мы уже профильтровали датасет. Можем ли мы там сразу замапить айдишники и подавать interactions в SequenceTaskConverter в новых внутренних айдишниках? Можем, это будет удобно читать. И SequenceTaskConverter вообще не должен думать о маппинге. Пусть делает только свою работу - формирует x и y из поданных интеракций.
`self.item_id_encoder` ты в коде класса `SequenceTaskConverter` ни разу не используешь - значит он тут не нужен. Подготовленный IdMap можно сохранить в атрибут `SasRecRecommenderModel`. Только не `self.item_id_encoder` а сразу `self.item_id_map`, чтобы не плодить термины в коде фреймворка

File: rectools/models/sasrec.py [hunk 4]
```diff
@@ -96,39 +101,33 @@ def fit(
             session_maxlen=self.session_maxlen,
             dropout_rate=self.dropout_rate,
             use_pos_emb=self.use_pos_emb,
-            n_items=len(self.processor.item_id_encoder) if self.processor.item_id_encoder is not None else -1,
```
Comment (blondered) [inline rectools/models/sasrec.py:99]: А он бывает None по твоему коду?

File: rectools/models/sasrec.py [hunk 5]
```diff
@@ -140,46 +139,59 @@ def recommend(
...
-    sessions: List[List[str]]
-    weights: List[List[float]]
+class SASRecModelInput:
+    """
+    sessions (torch.LongTensor): [batch_size, maxlen] user history
+    item_model_input (Tuple[torch.LongTensor, torch.LongTensor]): item features input to construct
+    """
+
+    sessions: torch.LongTensor
+    item_model_input: torch.LongTensor
```
Comment (blondered) [inline rectools/models/sasrec.py]: Давай дропать item_model_input везде. Это всегда одно и то же

File: rectools/models/sasrec.py [hunk 6]
```diff
@@ -641,69 +537,51 @@ def predict(
...
 
         epoch_start_idx = 1
 
         # ce_criterion = torch.nn.CrossEntropyLoss()
         # https://github.com/NVIDIA/pix2pixHD/issues/9 how could an old bug appear again...
 
-        iteration = 0
         try:
             for epoch in range(epoch_start_idx, self.epochs + 1):
                 logger.info("training epoch %s", epoch)
```
Comment (blondered) [inline rectools/models/sasrec.py:565]: Раз уж мы дропаем item_model_input, давай здесь сразу пофиксим:
В текущем варианте кода мы много раз прогоняем ItemModeldemb для всего каталога (на каждый батч заново считаем полные эмбеды всего каталога, чтобы их домножить на эмбеды сессий и получить логиты).
Кажется, что за эпоху это надо сделать всего один раз.

У SasRec давай сделаем отдельный метод, который считает эмбеды для всего каталога (результат ItemModeldemb.forward для всех айдишников). Отдельный - который считает эмбеды сессий. И потом уже результаты обоих методов перемножаем для получения логитов в методе forward.

Давай перед переходом в train_dataloader получим эмбеды всего каталога и затем будем передавать результат в self.train_step вместе с x, y и w.

Если непонятно - пиши, созвонимся. Эту доработку важно проверить отдельным коммитом и очень чётко проверить, что метрики не упали, а скорость стала быстрее (может и не супер значильно)

File: rectools/models/sasrec.py [hunk 7]
```diff
@@ -713,14 +591,13 @@ def fit(
...
 
         except KeyboardInterrupt:
             logger.info("training interrupted")
 
-    def train_step(self, x, y, w, iteration):
+    def train_step(self, x: torch.Tensor, y: torch.Tensor, w: torch.Tensor) -> None:
+        """TODO"""
         self.optimizer.zero_grad()
         logits = self.model(x)
         loss = self.loss_func(logits.transpose(1, 2), y)  # loss expects logits in form of [N, C, D1]
```
Comment (blondered) [inline rectools/models/sasrec.py]: [N, C, D1] ломает мне мозг каждый раз, когда я это вижу)
давай тут напишем супер-подробно:
```
# We are using CrossEntropyLoss with a multi-dimensional case
# Logits must be passed in form of [batch_size, n_classes, session_position] where session_position is one extra dimension
# Target label indexes must be passed in a form of [batch_size, session_position] ( `0` index for "PAD" ix excluded from loss)
# Loss output will have a shape of [batch_size, session_position] and will have zeros for every `0` target label
```

File: rectools/models/sasrec.py [hunk 8]
```diff
@@ -37,52 +38,56 @@ def __init__(
         use_pos_emb: bool = True,
         loss: str = "bce",
```
Comment (blondered) [inline rectools/models/sasrec.py]: bce лосс не поддерживается в коде. давай по умолчанию "sm_ce". Только назовём его "softmax"

File: rectools/models/sasrec.py [hunk 9]
```diff
@@ -731,73 +608,36 @@ def train_step(self, x, y, w, iteration):
...
-        else:
-            raise Exception(f"loss {self.loss} is not supported")
+    def _init_optimizers(self) -> torch.optim.Adam:
+        optimizer = torch.optim.Adam(self.model.parameters(), lr=self.lr, betas=(0.9, 0.98))
+        return optimizer
 
-        logger.info("used %s loss", self.loss)
+    def _init_loss_func(self, loss: str) -> Union[nn.BCEWithLogitsLoss, nn.CrossEntropyLoss]:
 
+        if loss == "bce":
```
Comment (blondered) [inline rectools/models/sasrec.py]: Дропаем. Нет поддержки в коде

File: rectools/models/sasrec.py [hunk 10]
```diff
@@ -64,25 +67,24 @@ def __init__(
...
+        user_item_interactions = user_item_interactions[(user_item_interactions[Columns.User].isin(users.index))]
 
         user_item_interactions = (
             user_item_interactions.sort_values(Columns.Datetime).groupby(Columns.User).tail(self.session_maxlen + 1)
         )
 
-        self.task_converter = SequenceTaskConverter(self.session_maxlen, user_item_interactions)
-        user_item_interactions[Columns.Item] = self.task_converter.item_id_encoder.convert_to_internal(
+        items = user_item_interactions[Columns.Item].drop_duplicates().sort_values()
+        self.item_id_map = self.item_id_map.add_ids(items)
```
Comment (blondered) [inline rectools/models/sasrec.py:84]: Мы в либе для всех моделей делаем одну стратегию: при повторном вызове fit на тех же или новых данных модель должна переобучиться с нуля. Потому в `_fit` нужно обязательно заново сделать `self.item_id_map = IdMap.from_values("PAD")` и только потом `add_ids`

File: rectools/models/sasrec.py [hunk 11]
```diff
@@ -64,25 +67,24 @@ def __init__(
...
-        user_item_interactions = user_item_interactions[(user_item_interactions[Columns.User].isin(users))]
+        user_item_interactions = user_item_interactions[(user_item_interactions[Columns.User].isin(users.index))]
 
         user_item_interactions = (
             user_item_interactions.sort_values(Columns.Datetime).groupby(Columns.User).tail(self.session_maxlen + 1)
         )
 
-        self.task_converter = SequenceTaskConverter(self.session_maxlen, user_item_interactions)
-        user_item_interactions[Columns.Item] = self.task_converter.item_id_encoder.convert_to_internal(
+        items = user_item_interactions[Columns.Item].drop_duplicates().sort_values()
```
Comment (blondered) [inline rectools/models/sasrec.py]: `np.unique(user_item_interactions[Columns.Item])` будет быстрее

File: rectools/models/sasrec.py [hunk 12]
```diff
@@ -506,8 +483,9 @@ def forward(
 
     def predict(
         self,
-        x: Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]],
+        sessions: torch.Tensor,
         item_indices: torch.Tensor,
+        item_model_input: torch.Tensor,
```
Comment (blondered) [inline rectools/models/sasrec.py]: Можно дропнуть

File: rectools/models/sasrec.py [hunk 13]
```diff
@@ -523,8 +501,6 @@ def predict(
             torch.Tensor: [batch_size, candidates_count] logits for each user
 
         """
-        sessions, item_model_input = x
-
         item_embs = self.item_model(item_model_input)
```
Comment (blondered) [inline rectools/models/sasrec.py]: Мы здесь подаём просто полный каталог.
`ItemNet` эта информация не нужна, она и так знает, что такое полный каталог.
Давай в ItemNet просто реализуем метод `get_all_embeddings` и здесь будем его вызывать.

Позже, когда ещё будем оптимизировать код в плане скорости, возможно сделаем метод `get_embeddins`, который на вход будет принимать список индексов айтемов и считать эмбеды только для них.

File: rectools/models/sasrec.py [hunk 14]
```diff
@@ -576,6 +544,7 @@ def fit(
 
         self.xavier_normal_init(self.model)
         self.model.train()  # enable model training
+        item_model_input = torch.LongTensor(list(range(self.model.n_items))).to(self.device)
```
Comment (blondered) [inline rectools/models/sasrec.py]: дропаем

File: rectools/models/sasrec.py [hunk 15]
```diff
@@ -679,14 +656,16 @@ def recommend(
...
 
-        user_item_interactions[Columns.Item] = self.task_converter.item_id_encoder.convert_to_internal(
+        user_item_interactions[Columns.Item] = self.item_id_map.convert_to_internal(
             user_item_interactions[Columns.Item]
         )
-        item_features = pd.Series(self.task_converter.item_id_encoder.convert_to_internal(item_features))
+        item_features = pd.Series(self.item_id_map.convert_to_internal(item_features))
+        items = item_features.sort_values().to_list()
+        item_model_input = np.concatenate([np.zeros_like(items[:1]), items], axis=0)
+        item_model_input = torch.LongTensor(item_model_input)
```
Comment (blondered) [inline rectools/models/sasrec.py]: дропаем item_model_input

**Review by blondered — APPROVED** [review]
